File: java/src/main/java/com/example/design_pattern/Factory.java
package com.example.design_pattern;

public class Factory {
  private static class Point {

    public enum Type {
      CARTESIAN,
      POLAR;
    }

    private double x;
    private double y;
    private Type type;

    public Point(double x, double y) {
      this.x = x;
      this.y = y;
      this.type = Type.CARTESIAN;
    }

    public Point(double a, double b, Type type) {
      if (type == Type.POLAR) {
        this.x = a * Math.cos(b);
        this.y = a * Math.sin(b);
        this.type = type;
        return;
      }

      this.x = a;
      this.y = b;
      this.type = type;
    }
  }

  private static class FactoryMethod {
    private static class Point {
      private double x;
      private double y;

      private Point(double x, double y) {
        this.x = x;
        this.y = y;
      }

      public static Point newCartesianPoint(double x, double y) {
        return new Point(x, y);
      }

      public static Point newPolarPoint(double r, double theta) {
        double x = r * Math.cos(theta);
        double y = r * Math.sin(theta);
        return new Point(x, y);
      }
    }

    public static void main() {
      Point p1 = Point.newCartesianPoint(1.0, 2.0);
      Point p2 = Point.newCartesianPoint(2.0, 3.0);
    }
  }

  private static class FactoryClass {
    private static class Point {
      private double x;
      private double y;

      protected Point(double x, double y) {
        this.x = x;
        this.y = y;
      }
    }

    private static class PointFactory {

      private PointFactory() {}

      public static Point newCartesian(double x, double y) {
        return new Point(x, y);
      }

      public static Point newPolar(double r, double theta) {
        double x = r * Math.cos(theta);
        double y = r * Math.sin(theta);

        return new Point(x, y);
      }
    }

    public static void main() {

      Point p1 = PointFactory.newCartesian(1.0, 2.0);
      Point p2 = PointFactory.newPolar(1.0, 2.0);
    }
  }

  private static class InnerFactory {
    private static class Point {
      private double x;
      private double y;

      public static class PointFactory {
        private PointFactory() {}

        public static Point newCartesian(double x, double y) {
          return new Point(x, y);
        }

        public static Point newPolar(double r, double theta) {
          double x = r * Math.cos(theta);
          double y = r * Math.sin(theta);

          return new Point(x, y);
        }
      }

      private Point(double x, double y) {
        this.x = x;
        this.y = y;
      }
    }

    public static void main() {
      Point p1 = Point.PointFactory.newCartesian(1.0, 2.0);
      Point p2 = Point.PointFactory.newPolar(1.0, 2.0);
    }
  }

  private static class HierarchicalFactory {

    private static interface Transport {
      public void deliver();
    }

    private static class Truck implements Transport {
      public void deliver() {}
    }

    private static class Ship implements Transport {
      public void deliver() {}
    }

    private static class Logistics {
      void someOperation() {
        Transport t = createTransport();
        t.deliver();
      }

      Transport createTransport() {
        return null;
      }
    }

    private static class RoadLogistics extends Logistics {
      Transport createTransport() {
        return new Truck();
      }
    }

    private static class SeaLogistics extends Logistics {
      Transport createTransport() {
        return new Ship();
      }
    }

    public static void main() {
      RoadLogistics rl = new RoadLogistics();
      SeaLogistics sl = new SeaLogistics();

      rl.someOperation();
      sl.someOperation();
    }
  }

  public static void main(String... args) {}
}

File: c++/design_pattern/Factory.cpp
#include <cmath>
#include <initializer_list>
#include <iostream>
#include <map>
#include <string>

using namespace std;

namespace ProblemCase {

enum class PointType { cartesian, polar };

struct Point {
  float x, y;
  Point(const float x, const float y) : x(x), y(y) {}  // 1. Cartisean 좌표계
  // Point(const float r, const float theta): x(r*cos(theta)),
  // y(r*sin(theta)){};  // 2. Polar 좌표계 :  (Error발생) 구분 불가
  Point(const float a, const float b, PointType type = PointType::cartesian) {
    if (type == PointType::cartesian) {
      x = a;
      y = b;
    } else {
      x = a * cos(b);
      y = a * sin(b);
    }
  }
};

};  // namespace ProblemCase

/* 1. 팩토리 메서드 */
/* - 충돌이 일어나는 생성자는 숨겨놓고 메서드로 생성자를 구현*/

namespace FactoryMethod {

struct Point {
  float x, y;

 protected:
  Point(const float x, const float y) : x(x), y(y) {};

 public:
  static Point NewCartesian(float x, float y) { return {x, y}; }
  static Point NewPolar(float r, float theta) {
    return {r * cos(theta), r * sin(theta)};
  }
};

void do_factory_method() { auto p = Point::NewPolar(5, 3.14); }

};  // namespace FactoryMethod
/* 2. 팩토리 */
/* - 팩토리 클래스를 생성*/

namespace Factory {

struct Point {
  float x, y;
  friend class PointFactory;

 private:
  Point(const float x, const float y) : x(x), y(y) {}
};

struct PointFactory {
  static Point NewCartesian(float x, float y) { return Point{x, y}; }

  static Point NewPolar(float r, float theta) {
    return Point{r * cos(theta), r * sin(theta)};
  }
};

void do_factory() { auto p = PointFactory::NewPolar(5, 3.14); }
};  // namespace Factory

/* 3. 내부 팩터리 */
/* - 내부 팩터리는 생성할 타입의 내부 클래스로 존재하는 간단한 팩터리 */

namespace InnerFactory {
struct Point {
 private:
  Point(float x, float y) : x(x), y(y) {};
  struct PointFactory {
   private:
    PointFactory() {}

   public:
    static Point NewCartesian(float x, float y) { return {x, y}; }
    static Point NewPolar(float r, float theta) {
      return {r * cos(theta), r * sin(theta)};
    }
  };

 public:
  float x, y;
  static PointFactory Factory;
};

void do_inner_factory() { auto p = Point::Factory.NewPolar(5, 3.14); }
}  // namespace InnerFactory

/* 4. 계층 팩터리 */
// 생성 객체 타입이 증가해도 클라이언트 코드를 수정할 필요가 없음

namespace HierarchicalFactory {

struct RoadLogistics;
struct SeaLogistics;

struct Transport {
  virtual void deliver() = 0;
};

struct Logistics {
  void someOperation();
  Transport* create_transport() { return nullptr; };
};

struct Truck : Transport {
  void deliver() {}
};

struct Ship : Transport {
  void deliver() {}
};

struct RoadLogistics : public Logistics {
  Transport* create_transport() { return new Truck(); }
};

struct SeaLogistics : public Logistics {
  Transport* create_transport() { return new Ship(); }
};

void Logistics::someOperation() {
  // ...
  Transport* t = create_transport();
  t->deliver();
  // ...
}

void do_hierarchical_factory() {
  Logistics l;
  RoadLogistics rl;
  SeaLogistics sl;

  l.someOperation();
  rl.someOperation();
  sl.someOperation();
}
};  // namespace HierarchicalFactory

int main(void) {
  FactoryMethod::do_factory_method();
  Factory::do_factory();
  InnerFactory::do_inner_factory();
  HierarchicalFactory::do_hierarchical_factory();

  return 0;
}